sources/packages/backend/apps/workers/src/services/disbursement-schedule/disbursement-schedule.service.ts

```diff
@@ -53,93 +106,179 @@ export class DisbursementScheduleService extends RecordDataModelService<Disburse
-          new Date(
-            previousSignedDisbursement.studentAssessment.offering.studyEndDate,
-          ),
-          // Start date of the offering of the current application is considered the end date.
-          new Date(firstDisbursement.studentAssessment.offering.studyStartDate),
-        );
-      }
+      // Reuse the MSFAA that is still valid and avoid creating a new one, even if it not signed yet.
+      return existingValidMSFAANumber.id;
+    }
 
-      if (hasValidMSFAANumber) {
-        // Reuse the MSFAA number.
-        msfaaNumberId = previousSignedDisbursement.msfaaNumber.id;
-      } else {
-        // Create a new MSFAA number in case the previous one is no longer valid.
-        const newMSFAANumber =
-          await this.msfaaNumberSharedService.createMSFAANumber(
-            applicationId,
-            systemUser.id,
-          );
-        msfaaNumberId = newMSFAANumber.id;
-      }
+    // Create a new MSFAA number in case no MSFAA is found or no longer valid.
+    const newMSFAANumber =
+      await this.msfaaNumberSharedService.createMSFAANumber(
+        applicationId,
+        auditUserId,
+      );
+    return newMSFAANumber.id;
+  }
+
+  /**
+   * Gets the student valid MSFAA number signed and the latest SFAS application study end date.
+   * @param studentId student id.
+   * @param offeringIntensity offering intensity.
+   * @returns SFAS signed MSFAA which is valid.
+   */
+  private async checkSFASSignedMSFAA(
+    studentId: number,
+    offeringIntensity: OfferingIntensity,
+  ): Promise<SFASSignedMSFAA> {
+    // Checks if there is a MSFAA number that could be considered valid from SFAS.
+    if (offeringIntensity === OfferingIntensity.fullTime) {
+      return await this.sfasApplicationService.getIndividualFullTimeApplicationByIndividualId(
+        studentId,
+      );
     }
-    const msfaaNumber = { id: msfaaNumberId } as MSFAANumber;
+    return await this.sfasPartTimeApplicationsService.getIndividualPartTimeApplicationByIndividualId(
+      studentId,
+    );
+  }
 
-    // Associate all the disbursements of the given assessment with MSFAA.
-    disbursements.forEach((disbursement) => {
-      disbursement.msfaaNumber = msfaaNumber;
-      disbursement.modifier = systemUser;
-    });
-    await this.repo.save(disbursements);
+  /**
+   * Checks if the MSFAA number is still valid.
+   * If the study period end date of the previously signed MSFAA is less than 2 years
+   * when compared to current study period start date, then MSFAA is considered to be valid.
+   * @param previousSignedDisbursement previously signed disbursement.
+   * @param firstDisbursement first disbursement of the application.
+   * @returns whether the MSFAA Number is valid.
+   */
+  private async isMSFAANumberStillValid(
+    previousSignedDisbursement: DisbursementSchedule,
+    firstDisbursement: DisbursementSchedule,
```

Comment:
Please be more precise about the parameters required for this method.
For instance `firstDisbursement` is provided to use the `applicationId`, `studentId`, and as a parameter to `isMSFAANumberStillValid` where the entire object is not needed.

Reply:
Parameter names in the comments must be updated.